r1cs/gadgets/std/src/fields/fp2.rs

```diff
@@ -180,131 +228,231 @@ impl<P: Fp2Parameters<Fp = ConstraintF>, ConstraintF: PrimeField + SquareRootFie
+        let a0_plus_non_residue_c1 = self
+            .c0
+            .add(cs.ns(|| "a0 + non_residue * a1"), &self.c1)?;
         let one_plus_non_residue_v0 = v0.mul_by_constant(
-            cs.ns(|| "1 + non_residue * v0"),
+            cs.ns(|| "(1 + non_residue) * v0"),
             &(P::Fp::one() + &P::NONRESIDUE),
         )?;
-
+        // constraint 2
         self.c0 = a0_plus_a1
             .mul(
                 cs.ns(|| "(a0 + a1) * (a0 + non_residue * a1)"),
                 &a0_plus_non_residue_c1,
             )?
-            .sub(cs.ns(|| "- (1 + non_residue) v0"), &one_plus_non_residue_v0)?;
+            .sub(cs.ns(|| "- (1 + non_residue) * v0"), &one_plus_non_residue_v0)?;
 
+        // no extra constraint, works directly on LC
         v0.double_in_place(cs.ns(|| "2v0"))?;
         self.c1 = v0;
 
         Ok(self)
     }
 
+
+    // now that the patched code does not save any constraints:
+    // why not replace Karatsuba code by simple call to mul_equals?
+    #[inline]
+    fn inverse<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Self, SynthesisError> {
+        let inverse = Self::alloc(&mut cs.ns(|| "alloc inverse"), || {
+            self.get_value().and_then(|val| val.inverse()).get()
+        })?;
+
+        // Karatsuba multiplication for Fp2 with the inverse:
+        //     v0 = A.c0 * B.c0,
+        //     v1 = A.c1 * B.c1,
+        //      1 = v0 + non_residue * v1,
+        //      0 = result.c1 = (A.c0 + A.c1) * (B.c0 + B.c1) - v0 - v1.
+        // Enforced with 3 constraints (substituting v0 by v1)
+        //    (1)  A.c1 * B.c1 = v1,
+        //    (2) (A.c0 + A.c1) * (B.c0 + B.c1) =  1 + (1 - non_residue) * v1
+        //                                      = 1 - non_residue * v1 + v1
+        //    (3)  A.c0 * B.c0 = 1 - non_residue * v1,
+        // Reference:
+        // "Multiplication and Squaring on Pairing-Friendly Fields"
+        // Devegili, OhEigeartaigh, Scott, Dahab
+
+        // Constraint 1
+        let v1 = self.c1.mul(cs.ns(|| "inv_constraint_1"),
+                             &inverse.c1)?;
+
+        // Constraint 2
+        let one = P::Fp::one();
+        let a0_plus_a1 = self.c0.add(cs.ns(|| "a0 + a1"), &self.c1)?;
+        let b0_plus_b1 = inverse.c0.add(cs.ns(|| "b0 + b1"), &inverse.c1)?;
+
+        let rhs =
+            v1.mul_by_constant(
+                cs.ns(|| "(1 - nonresidue) * v1"),
+                &(one - &P::NONRESIDUE)
+            )?
+                .add_constant(
+                    cs.ns(|| "add one"),
+                    &one)?;
+
+        a0_plus_a1.mul_equals(cs.ns(|| "inv_constraint_2"), &b0_plus_b1, &rhs)?;
+
+        // Constraint 3
+        let rhs2 = rhs.sub(cs.ns(|| " 1 - nonresidue * v1"), &v1)?;
+        self.c0.mul_equals(cs.ns(||"inv_constraint_3"),&inverse.c0, &rhs2)?;
+
+        Ok(inverse)
+    }
+
+    // does not save any constraint over default implementation?
```

Comment:
Yes it does. Multiplication + enforce equal would've cost you 5 constraints, while with this gadget will be only 3